### PR TITLE
Add show_read_posts filter. Fixes #1561

### LINKED ISSUE #1561: Mark posts as seen / hide read posts option
### Describe the solution you'd like

A button that marks posts as seen and removes them from the front page.

### Describe alternatives you've considered

An option that makes lemmy mark posts as seen if they are opened.

### PR DESCRIPTION
**Pull request by dessalines**

### CODE REVIEW

**Comment by dessalines** [comment]
I built and tested the front end for this, it works.

File: crates/api/src/post.rs [hunk 1]
```diff
@@ -6,6 +6,7 @@ use lemmy_api_common::{
   check_downvotes_enabled,
```
Comment (dessalines) [inline crates/api/src/post.rs:6]: I auto-mark liked and saved posts as read.

File: crates/api_crud/src/post/create.rs [hunk 2]
```diff
@@ -92,6 +100,9 @@ impl PerformCrud for CreatePost {
       return Err(ApiError::err("couldnt_like_post").into());
     }
 
+    // Mark the post as read
+    mark_post_as_read(person_id, post_id, context.pool()).await?;
```
Comment (dessalines) [inline crates/api_crud/src/post/create.rs:104]: Created too.

File: crates/api_crud/src/post/read.rs [hunk 3]
```diff
@@ -43,6 +45,11 @@ impl PerformCrud for GetPost {
     .await?
     .map_err(|_| ApiError::err("couldnt_find_post"))?;
 
+    // Mark the post as read
+    if let Some(person_id) = person_id {
+      mark_post_as_read(person_id, id, context.pool()).await?;
+    }
```
Comment (dessalines) [inline crates/api_crud/src/post/read.rs:51]: And indivdually read posts, IE `GetPost`

File: migrations/2021-04-24-174047_add_show_read_post_setting/up.sql [hunk 4]
```diff
@@ -0,0 +1 @@
+alter table local_user add column show_read_posts boolean default true not null;
```
Comment (dessalines) [inline migrations/2021-04-24-174047_add_show_read_post_setting/up.sql:1]: I could've done "hide read", but all the other ones are named `show`, so this seems more consistent.

File: crates/api_common/src/lib.rs [hunk 5]
```diff
@@ -252,12 +253,34 @@ pub fn user_show_nsfw(local_user_view: &Option<LocalUserView>) -> bool {
   }
 }
 
+/// A helper method for showing read posts
+pub fn user_show_read_posts(local_user_view: &Option<LocalUserView>) -> bool {
+  match local_user_view {
+    Some(uv) => uv.to_owned().local_user.show_read_posts,
+    None => true,
+  }
```
Comment (Nutomic) [inline crates/api_common/src/lib.rs:261]: This could also be a simple map.

Comment (dessalines) [reply crates/api_common/src/lib.rs:261]: I think doing it that way, i'll be able to get rid of these helper functions entirely. I'll do that shortly.

**Review by Nutomic — APPROVED** [review]

**Pull request merged**